[MCA] Enable customization of individual instructions

r-belenov opened the pull request:
Currently MCA takes instruction properties from scheduling model. However, some instructions may execute differently depending on external factors - for example, latency of memory instructions may vary differently depending on whether the load comes from L1 cache, L2 or DRAM. While MCA as a static analysis tool cannot model such differences (and currently takes some static decision, e.g. all memory ops are treated as L1 accesses), it makes sense to allow manual modification of instruction properties to model different behavior (e.g. sensitivity of code performance to cache misses in particular load instruction). This patch addresses this need.

The library modification is intentionally generic - arbitrary modifications to InstrDesc are allowed. The tool support is currently limited to changing instruction latencies (single number applies to all output arguments and MaxLatency) via coments in the input assembler code; the format is the like this:

add (%eax), eax // LLVM-MCA-LATENCY:100

Users of MCA library can already make additional customizations; command line tool can be extended in the future.

Note that InstructionView currently shows per-instruction information according to scheduling model and is not affected by this change.

See https://github.com/llvm/llvm-project/issues/133429 for additional clarifications (including explanation why existing customization mechanisms do not provide required functionality)

r-belenov committed:
Added interface to customize individual instructions and possibility to change latency via special comments

* Added customizations to header

* Added customization to implementation

* Use default small vector size for custom instructions

* Typo fixed

* Use std::vector for custom descriptors

* Revert to SmallVector

* Erroneous dereference removed

* Introduced annotated instructions

* Updated getter method

* Updated getter usage

* Fixed type usage

* Revert to MCInst for regions

* Keep annotations in separate map

* Removed usage of annotated instruction

* Use std::optional

* Add setter/getter for explicit latency

* Set instruction latency according to annotation

* Simplify default return

* Revert to explicit default value

* Add missing brace

* Added missing field name

* Typo fixed

* Added missing namespace tags

* Whitespace removed

* Index annotations by smloc

* Search latency by SMLoc

* Use pointer as a map key

* Missing const added

* Added optional annotation to MCStreamerWrapper

* Added interface to annotate instructions

* Move annotations to CodeRegons

* Get explicit latency from Regions

* Erroneous dereference removed

* Add interface to pass latency to MCStreamerWrapper

* Fixed incorrect code placement

* Make Streamer available to CommentConsumer

* Move MCStreamerWrapper definition to make it visible in comment consumers

* Parse latency comment

* Fix drop_front usage

* Fix whitespace skipping on latency annotation

* Added test for latency annotation

* Additional test for latency annotations

* Use load in test instruction

* Use load in test instruction

* Added instruction builder customization

* Added customized instruction creation

* Pass builder to custom creation functor

* Added header with InstrBuilder declaration

* Typo fixed

* Added base for instruction customizations test

* Added missing result object

* Hardcode name for summary view

* Explicitly model single instruction

* Add actual custom latency test

* Typo fixed

* Pass latency to lambda body

* Adding missing const

* Specify lambda return type

* Erroneous argument removed

* Removed unnecessary lambda return type

* Code cleanup

* Code cleanup

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Fix formatting

r-belenov committed:
Merge branch 'main' into mca-instruction-customizer

mshockwave reviewed (commented):
Could you add another new callback in `InstrumentManager` to implement the customizer, and move the parsing logics of `LLVM-MCA-LATENCY` from CodeRegionGenerator.* to InstrumentManager (either putting it in X86's custom behavior or in InstrumentManager's base implementation) because InstrumentManager was created exactly for supporting custom annotations like `LLVM-MCA-LATENCY`.

r-belenov committed:
Merge branch 'main' into mca-instruction-customizer

r-belenov committed:
Using function_ref to pass function arguments

* Switch to function_ref

* Switch to function_ref

* Check pipeline execution status

* Switch to function_ref

r-belenov commented on llvm/include/llvm/MCA/InstrBuilder.h:
```diff
@@ -87,7 +92,8 @@ class InstrBuilder {
   Expected<const InstrDesc &>
-  createInstrDescImpl(const MCInst &MCI, const SmallVector<Instrument *> &IVec);
+  createInstrDescImpl(const MCInst &MCI, const SmallVector<Instrument *> &IVec,
+                      std::function<void(InstrDesc &)> Customizer = {});
```
Done

r-belenov committed:
Use InstrumentManager to customize instructions

Extended InstrumentManager to allow InstrDesc modification with instruments; added common instrument that can modify instruction latencies.

r-belenov committed:
Fix formatting

Fixed formatting issues

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -143,19 +146,27 @@ class LLVM_ABI InstrumentManager {
   const MCSubtargetInfo &STI;
   const MCInstrInfo &MCII;
+  bool EnableDefaults;
+  std::unique_ptr<InstrumentManager> TargetIM;
```
InstrumentManager is already target-specific, why do you need TargetIM here?

mshockwave replied:
> I'd like to have InstrDesc customization generic (e.g. latency change code is generic, other things I do locally also do not require target details) and coexist with target-specific instrumentation.

I think my concern here is that this additional layer of pointer indirection could be avoid if we could simply put this latency updating logics in the base IM and called it from the target-specific, derived IM when needed, as I described in the sibling comment

r-belenov replied:
Ok, now the customization in InstrumentManager class and can be reused.

mshockwave commented on llvm/lib/MCA/CustomBehaviour.cpp:
```diff
@@ -42,19 +43,91 @@ CustomBehaviour::getEndViews(llvm::MCInstPrinter &IP,
   return std::vector<std::unique_ptr<View>>();
 }
 
+class CustomInstrument : public Instrument {
```
I think this can be solved in a much simpler way: given a line "LLVM-MCA-LATENCY: 100", the default `Instrument` could already provide you `Desc` and `Data` of "LLVM-MCA-LATENCY:" and "100" respectively. So I don't think you need the `CustomInstrument` here. 
And then for `supportsInstrumentType`, `canCustomize` and `customize` -- you could just put the implementation directly in the function, and call this base implementation for the target-specific override, if a target overrides it. Such that you don't need the slightly confusing `TargetIM` indirection.

r-belenov replied:
> given a line "LLVM-MCA-LATENCY: 100", the default Instrument could already provide you Desc and Data of "LLVM-MCA-LATENCY:" and "100" respectively. So I don't think you need the CustomInstrument here.
Ok, I can move the logic to base Instrument implementation
> Such that you don't need the slightly confusing TargetIM indirection.
Without TargetIM generic customization will be disabled in case target-specific IM is used (since it will override createInstrument), unless all target IMs call base class createInstrument for unrecognized types.

r-belenov replied:
Also, supportsIntrumentType() is normally overridden in target-specific IMs, while we need to check for both common and target specific instruments.

r-belenov replied:
Ok, I'll change the code to leave the logic in the base IM only without TargetIM indirection.

r-belenov replied:
PS Pushed the changes that remove targetIM

mshockwave reviewed (commented):
I think your current solution could be (greatly) simplified, please see my inline comment

r-belenov committed:
Move latency customization logic to base Instrument

r-belenov committed:
Simplify IM creation

r-belenov committed:
Fix formatting

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -123,13 +126,41 @@ class Instrument {
   virtual ~Instrument() = default;
 
+  virtual bool canCustomize() const { return bool(Latency); }
+  virtual void customize(InstrDesc &ID) const {
```
I would prefer not adding this virtual function here and implement the customization logics directly in `InstrumentManager::customize` instead.
With such approach, we can also avoid storing Latency in this class.

r-belenov replied:
Latency must be stored in the instrument since it may be (and normally is) different in different instrumentation regions.

r-belenov replied:
Ping - any additional comments?

r-belenov committed:
Remove target IM from base class

r-belenov committed:
Fix formatting

* Update CustomBehaviour.h

* Update CustomBehaviour.cpp

* Update llvm-mca.cpp

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -123,13 +123,41 @@ class Instrument {
+  virtual void customize(InstrDesc &ID) const {
+    if (Latency) {
+      for (auto &W : ID.Writes)
+        W.Latency = *Latency;
```
@adibiagio pointed this out in our offline discussions: an instruction may perform multiple writes, is there any way we could designate which write to apply this latency customization? It doesn't have to be in this patch, but a TODO comment would be nice.

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -143,19 +171,21 @@ class LLVM_ABI InstrumentManager {
 protected:
   const MCSubtargetInfo &STI;
   const MCInstrInfo &MCII;
+  bool EnableInstruments;
```
I was wrapping my head around the purpose of this flag, since it's always true among all the instances in the codebase. But I think there is a nice way to move forward: historically, we use `-disable-im` in llvm-mca to instantiate a default `InstrumentManager` (instead of a target-specific one) -- but that was based on the fact that the default `InstrumentManager` do absolutely no instrumentation. And this patch is changing this behavior. So perhaps we could either: (1) having two flags -- one switching to a target-independent InstrumentManager like what `-disable-im` is doing right now, and the other being passed to the `EnableInstruments` here. (2) having just one flag to toggle the `EnableInstruments` here, and each target-specific InstrumentManager is expected to honor this `EnableInstruments`.

mshockwave reviewed (commented):
Personally I prefer to detach as many logics as possible away from the `Instrument` class: it's supposed to be a simple base class for targets to override. More importantly, I believe it's totally possible to do _all_ the latency comment parsing in `InstrumentManager` without making any change in the `Instrument` class -- for a given `Instrument`, parse the `Instrument::Data` when the `Instrument::Desc` is equal to "LLVM-MCA-LATENCY:" (you don't really need the `CustomInstrumentName`, parsing the raw Desc and Data should suffice in this case).  With such, we can also get rid of `Instrument::customize` and `Instrument::canCustomize`.

Another high-level comment: please add some documentations about this new feature in `docs/CommandGuide/llvm-mca.rst`

mshockwave commented:
> We cant just keep the strings and do the parsing in the callbacks called from InstrBuilder, but that looks too expensive for me (especially in case MCA is used as a library and latency already comes in numeric form from some external source, so no parsing is required).

I was indeed thinking about re-parsing every time, but you have a point here and I agree, so maybe we can restore to one of your previous approaches which create a new class derived from `Instrument` (it was called `CustomInstrument` IIRC, but we can have some more specific name like `LatencyInstrument`). That being said, I still don't think we should have the `Instrument::customize` and `Instrument::canCustomize` -- I believe we can conditionally static_cast to this new LatencyInstrument by the getDesc it returns.

r-belenov committed:
Use explicit latency instrument

Latency customization logic is moved to dedicated instrument.

r-belenov committed:
Fix formatting (#12)

* Update CustomBehaviour.h

* Update TestIncrementalMCA.cpp

r-belenov committed:
Removing redundant include

<functional> was required for initial approach with customization passed via std::function, now we use IM methods.

r-belenov committed:
Merge branch 'main' into mca-instruction-customizer

r-belenov committed:
Make latency test more consistent

Ensure the same value is used to set the latency and test the result

r-belenov committed:
Merge branch 'main' into mca-instruction-customizer

mshockwave commented on llvm/lib/MCA/CustomBehaviour.cpp:
```diff
@@ -42,8 +43,35 @@ CustomBehaviour::getEndViews(llvm::MCInstPrinter &IP,
+const llvm::StringRef LatencyInstrument::DESC_NAME = "LATENCY";
+
+bool InstrumentManager::supportsInstrumentType(StringRef Type) const {
+  return EnableInstruments && Type == LatencyInstrument::DESC_NAME;
```
assuming a target-specific InstrumentManager also wants to support latency instrument, then I don't think it can do something like this right?
``` c++
bool MyTargetInstrumentManager::supportsInstrumentType(StirngRef Type) const override {
  // target-specific logics
  return InstrumentManager::supportsInstrumentType(Type);
}
```
is there a more ergonomic way to use latency instrument (or any default instruments added in the future) from a target IM? Should we set `EnableInstruments` default to true?

r-belenov replied:
Agree, changed default EnablInsturments value

mshockwave commented:
> > That being said, I still don't think we should have the Instrument::customize and Instrument::canCustomize -- I believe we can conditionally static_cast to this new LatencyInstrument by the getDesc it returns.
> 
> We still need callbacks in InstrBuilder to check whether some instrument can customize InstrDesc (so we cannot reuse the description) and to do the customization, and latency change is not the only possible customization (though it is the only one in the current patch), so I don't like having the code in InstrumentManager that checks for specific instruments.

Instrument was originally invented as an opaque "handle" pass to InstrumentManager, and InstrumentManager does the actual works. Otherwise we would have a `Instrument::getSchedClassID` virtual function (because some instruments are used by `InstrumentManager::getSchedClassID`). For every new features added into `InstrumentManager` in the future, for instance `InstrumentManager::doFoo`,  if we also need to add a new virtual function counterpart into `Instrument`, like `Instrument::doFoo`, I don't think it'll scale well.

Other than that, I think most parts of this patch are coming into shape, thanks for pushing this work!

r-belenov committed:
Move customization logic to InstrumentManager

r-belenov commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -513,10 +513,14 @@ int main(int argc, char **argv) {
+      IM = std::make_unique<mca::InstrumentManager>(*STI, *MCII, true);
+    }
+  } else {
+    // If the -disable-cb flag is set then we use the default base class
```
Ok, fixed

r-belenov committed:
Fix formatting (#14)

r-belenov committed:
Merge branch 'main' into mca-instruction-customizer

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -134,6 +134,27 @@ class Instrument {
+  }
+
+  bool hasValue() const { return bool(Latency); }
+  unsigned getLatency() { return *Latency; }
```
I believe this should also be `const`

mshockwave commented on llvm/lib/MCA/CustomBehaviour.cpp:
```diff
@@ -42,8 +43,45 @@ CustomBehaviour::getEndViews(llvm::MCInstPrinter &IP,
+    if (I->getDesc() == LatencyInstrument::DESC_NAME) {
+      auto LatInst = static_cast<LatencyInstrument *>(I);
+      if (LatInst->hasValue()) {
+        auto Latency = LatInst->getLatency();
```
please spell out the type

mshockwave commented on llvm/unittests/tools/llvm-mca/X86/TestIncrementalMCA.cpp:
```diff
@@ -234,3 +234,27 @@ TEST_F(X86TestBase, TestVariantInstructionsSameAddress) {
+  MCIs.push_back(InstructionToAdd);
+  SmallVector<std::pair<StringRef, StringRef>> InstrDescs;
+  auto LatStr = std::to_string(ExplicitLatency);
+  InstrDescs.push_back(std::make_pair(StringRef("LATENCY"), StringRef(LatStr)));
```
I thought StringRef has implicit conversion from both `const char *` and std::string. Does it work if you remove the `StringRef(...)` here?

r-belenov replied:
I tried and had no issues with local build and test, but it led to UT failure in github CI. Note that conversion omission leads to creation of temporary pair (since make_pair construct pair<char*,string>) and it looks that it causes UB here.

r-belenov replied:
PS As far as I see, without explicit conversion StringRef with latency value references the content of temporary string in temporary pair which is destroyed immediately after this statement.

mshockwave commented on llvm/unittests/tools/llvm-mca/MCATestBase.h:
```diff
@@ -73,9 +74,11 @@ class MCATestBase : public ::testing::Test {
+  runBaselineMCA(json::Object &Result, ArrayRef<MCInst> Insts,
+                 ArrayRef<mca::View *> Views = {},
+                 const mca::PipelineOptions *PO = nullptr,
+                 SmallVector<std::pair<StringRef, StringRef>> Descs = {});
```
please use ArrayRef here: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h (see the "Note")

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -134,6 +134,27 @@ class Instrument {
+public:
+  static const llvm::StringRef DESC_NAME;
+  LatencyInstrument(StringRef Data) : Instrument(DESC_NAME, Data) {
+    // Skip spaces and tabs.
+    unsigned Position = Data.find_first_not_of(" \t");
+    if (Position >= Data.size())
+      // We reached the end of the comment. Bail out.
+      return;
+    Data = Data.drop_front(Position);
```
can we use `StringRef::trim` to replace this part of the code?

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -134,6 +134,27 @@ class Instrument {
+  std::optional<unsigned> Latency;
+
+public:
+  static const llvm::StringRef DESC_NAME;
```
drop "llvm::"

mshockwave commented on llvm/docs/CommandGuide/llvm-mca.rst:
```diff
@@ -383,6 +383,20 @@ that do not start with `LLVM-MCA-` are ignored by :program:`llvm-mca`.
+  mov (%edi), %eax
+  # LLVM-MCA-LATENCY
+
+It set the latency of mov instruction to 100. LLVM-MCA-LATENCY without
```
```suggestion
It sets the latency of mov instruction to 100. LLVM-MCA-LATENCY without
```

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -175,6 +198,13 @@ class LLVM_ABI InstrumentManager {
                                    const SmallVector<Instrument *> &IVec) const;
+
+  // Return true if instruments can modify instruction description
+  virtual bool canCustomize(const SmallVector<Instrument *> &IVec) const;
```
please use ArrayRef here

mshockwave commented on llvm/include/llvm/MCA/CustomBehaviour.h:
```diff
@@ -175,6 +198,13 @@ class LLVM_ABI InstrumentManager {
+  virtual bool canCustomize(const SmallVector<Instrument *> &IVec) const;
+
+  // Customize instruction description
+  virtual void customize(const SmallVector<Instrument *> &IVec,
```
ditto `ArrayRef` for IVec

mshockwave reviewed (approved):
LGTM w/ the condition that the last batch of minor comments are addressed.
Cheers!

r-belenov committed:
FIxed incorrect Enlish

Co-authored-by: Min-Yih Hsu <min@myhsu.dev>

r-belenov committed:
Addressed minor comments

r-belenov committed:
Fixed formatting

r-belenov committed:
Merge branch 'main' into mca-instruction-customizer

r-belenov committed:
Ensure test instruction can not be eliminated on dispatch stage

r-belenov committed:
Merge branch 'main' into mca-instruction-customizer

r-belenov committed:
Revert to explicit StringRef conversion to check the tests
